File: 05_dependency_inversion/dart/index-good.dart
class Store {
  Store(PaymentProcesador paymentProcesador) {
    this.paymentProcesador = paymentProcesador;
  }
  late PaymentProcesador paymentProcesador;

  purchaseBike(int quantity) {
    this.paymentProcesador.pay(200 * quantity);
  }

  purchaseHelmet(int quantity) {
    this.paymentProcesador.pay(15 * quantity);
  }
}

abstract class PaymentProcesador {
  void pay(int amountInDollars) {}
}

class StripePaymentProcesador implements PaymentProcesador {
  StripePaymentProcesador(dynamic user) {
    this.stripe = Stripe(user);
  }
  late Stripe stripe;
  pay(int amountInDollars) {
    this.stripe.makePayment(amountInDollars * 100);
  }
}

class Stripe {
  Stripe(dynamic user) {
    this.user = user;
  }
  dynamic user;

  makePayment(int amountInCents) {
    print(
      '${this.user} made payment of \$${amountInCents / 100} with Stripe',
    );
  }
}

class PaypalPaymentProcesador implements PaymentProcesador {
  PaypalPaymentProcesador(dynamic user) {
    this.user = user;
    this.paypal = Paypal();
  }
  dynamic user;
  late Paypal paypal;

  pay(int amountInDollars) {
    this.paypal.makePayment(this.user, amountInDollars);
  }
}

class Paypal {
  makePayment(dynamic user, int amountInDollars) {
    print('${user} made payment of \$${amountInDollars} with Paypal');
  }
}

main() {
  // Store store = Store(StripePaymentProcesador("John"));
  Store store = Store(PaypalPaymentProcesador("John"));
  store.purchaseBike(2);
  store.purchaseHelmet(2);
}

File: 05_dependency_inversion/typescript/index-good.ts
class Store {
  constructor(paymentProcesador: PaymentProcesador) {
    this.paymentProcesador = paymentProcesador;
  }
  paymentProcesador: PaymentProcesador;

  purchaseBike(quantity: number) {
    this.paymentProcesador.pay(200 * quantity);
  }
  purchaseHelmet(quantity: number) {
    this.paymentProcesador.pay(15 * quantity);
  }
}

abstract class PaymentProcesador {
  pay(amountInDollars: number): void {}
}

class StripePaymentProcesador implements PaymentProcesador {
  constructor(user: any) {
    this.stripe = new Stripe(user);
  }
  stripe: Stripe;
  pay(amountInDollars: number) {
    this.stripe.makePayment(amountInDollars * 100);
  }
}

class Stripe {
  constructor(user: any) {
    this.user = user;
  }
  user: any;

  makePayment(amountInCents: number) {
    console.log(
      `${this.user} made payment of \$${amountInCents / 100} with Stripe`,
    );
  }
}

class PaypalPaymentProcesador implements PaymentProcesador {
  constructor(user: any) {
    this.user = user;
    this.paypal = new Paypal();
  }
  user: any;
  paypal: Paypal;

  pay(amountInDollars: number) {
    this.paypal.makePayment(this.user, amountInDollars);
  }
}

class Paypal {
  makePayment(user: any, amountInDollars: number) {
    console.log(`${user} made payment of \$${amountInDollars} with Paypal`);
  }
}

const store = new Store(new PaypalPaymentProcesador("John"));
store.purchaseBike(2);
store.purchaseHelmet(2);
